### ISSUE TITLE
[JOSS Review]: Python packaging is required

### ISSUE THREAD

**Issue by matthewfeickert** [umami-hep/umami#6]
(Continuing from https://github.com/openjournals/joss-reviews/issues/5833#issuecomment-1990949331)

In the [Submitting a paper to JOSS docs](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort) it is noted that

> In addition, JOSS requires that software should be feature-complete (i.e., no half-baked solutions), packaged appropriately according to common community standards for the programming language being used (e.g., [Python](https://packaging.python.org/), [R](https://r-pkgs.org/index.html)), and designed for maintainable extension (not one-off modifications of existing tools).

which means that for Python packages they should be distributed via a package index (ideally PyPI, or conda-forge if there are complicated external binaries) and are appropriately packaged such that they can be installed with their runtime dependencies. Providing Linux container images in lieu of proper packaging isn't sufficient for JOSS publications (though they are welcome to provide as an addition!). For submissions like `umami` that are not designed to be more general libraries, the use motivation for distributing on a package index is lower and so won't be strongly recommended, however all JOSS submissions need to be packaged appropriately.

In short, this means that for libraries all the required runtime dependencies need to be included in the packaging metadata with known lower bounds (if you don't want to empirically determine these lower bounds it is okay to select reasonable known working versions) so the installer knows what versions are required. For a Python application, in addition to the library requirements, a `requirements.txt` or (preferably) a lock file (created by some well established tool like `pip-tools`, `conda-lock`, `poetry`, `uv` as there is [currently no formal Python lock file standard](https://discuss.python.org/t/lock-files-again-but-this-time-w-sdists/46593)) should also be provided.

(An excellent, and current, overview of packaging is provided by the [Scientific Python organization Development Guide](https://learn.scientific-python.org/development/) and as `umami` doesn't have compiled extensions the [simple packaging section](https://learn.scientific-python.org/development/guides/packaging-simple/) would be relevant.)

From e19017383d1385978d7926b036e284f7602fee30, `umami`'s build backend is `setuptools`

https://github.com/umami-hep/umami/blob/e19017383d1385978d7926b036e284f7602fee30/setup.py#L9-L11

https://github.com/umami-hep/umami/blob/e19017383d1385978d7926b036e284f7602fee30/pyproject.toml#L1-L2

though all [dependency metadata](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#declaring-required-dependency) is missing from [`setup.cfg`](https://github.com/umami-hep/umami/blob/e19017383d1385978d7926b036e284f7602fee30/setup.cfg).

The current [`umami` installation docs](https://umami.docs.cern.ch/setup/installation/#cloning-the-repository) note

https://github.com/umami-hep/umami/blob/e19017383d1385978d7926b036e284f7602fee30/docs/setup/installation.md?plain=1#L231-L237

though it is unclear what restrictions exist that would keep `tensorflow` from being installed into a clean virtual environment.

My recommendation for packaging things properly with the least amount of work is:

1. Establish from the source code what are the required runtime dependencies and provide them with lower bounds in your packaging metadata. If you wanted to take this as an opportunity to simplify your configuration files and move to just using `pyproject.toml` this would be a good opportunity, though that is not required, and you can keep `setup.cfg` if you want. Regardless, please update the build system metadata in `pyproject.toml` to provide build backend information

```toml
[build-system]
requires = ["setuptools>=61.0"]
build-backend = "setuptools.build_meta"
```

I would assume that most of these dependencies are already well summarized in

https://github.com/umami-hep/umami/blob/e19017383d1385978d7926b036e284f7602fee30/requirements.txt#L1-L16

though `tensorflow` is notably absent and so perhaps others are as well.

2. Update `requirements.txt` to provide correct application run requirements with deploy/run versions pinned (or optionally produce a lock file).

3. It seems that images from TensorFlow are used as base images for the Docker container images

https://github.com/umami-hep/umami/blob/e19017383d1385978d7926b036e284f7602fee30/.gitlab/workflow/.docker-gitlab-ci.yaml#L58-L59

https://github.com/umami-hep/umami/blob/e19017383d1385978d7926b036e284f7602fee30/docker/umamibase/Dockerfile#L17-L21

so if having `tensorflow` in a `requirements.txt` file is going to cause conflicts here, you can always do a precautionary clean with `sed`

```docker
RUN sed -i '/tensorflow/d' requirements.txt && python -m pip install -r requirements.txt
```

You might have more clever ideas, but that's a naive suggestion.

4. Update the installation docs to reflect that the dependencies are provided correctly.

This is a lot of text, but the work required here I think is quite minimal as I believe most of the information already exists. If the authors have questions about any of the above, please ask here. :+1:

**Comment by matthewfeickert** [comment]
:wave: @philippgadow I'm just checking in on this Issue for https://github.com/openjournals/joss-reviews/issues/5833 not to rush you and the team or be pushy, but to make it clear that I haven't forgotten about this review and abandoned it. If there are questions that I can provide any recommendations or further input on I am happy to do so.

**Comment by afroch** [comment]
Hi @matthewfeickert. Sorry for the long delay, but I'm currently looking into your comments from above. I adapted already your bullet points 3 and 4 but I have a question concerning your first and second bullet point:

1: When I understand this correctly, the TODO here is to update the `pyproject.toml` with the little code snippet you provided and add the requirements from the `requirements.txt` file to the `setup.cfg`. Is this correct? If so, I saw there is a possibility to do this dynamically. I guess this would be good?

2: Which packages do you talk about when writing "correct application run requirements with deploy/run versions pinned"? From my point, I think we pinned down all the different packages that we use with the version required. Could you clarify what you meant with that?

**Comment by matthewfeickert** [comment]
Sorry for _my_ long delay, @afroch, and the inconveniences it has caused.

> I adapted already your bullet points 3 and 4

Fantastic and thank you. Do you happen to have these on a branch of https://github.com/umami-hep/umami/ or a public fork that I can look at and give feedback on? Feel free to open PRs to address them and tag me there too.

> 1: When I understand this correctly, the TODO here is to update the pyproject.toml with the little code snippet you provided

Correct. The build-system metadata is incomplete at the moment

https://github.com/umami-hep/umami/blob/c7e7d164f554f07a20dcfbf760a47c828e638abc/pyproject.toml#L1-L2

and so not compliant with [PEP 518](https://peps.python.org/pep-0518/), which will be a problem in the future for installers. All that is required here is to correctly specify the `build-backend` metadata and update the `requires`. So you can just update the `[build-system]` to be

```toml
[build-system]
requires = ["setuptools>=61.0"]
build-backend = "setuptools.build_meta"
```

c.f. The [Scientific Python library development guide on simple packaging](https://learn.scientific-python.org/development/guides/packaging-simple/#pyprojecttoml-build-system) for more details.

> and add the requirements from the requirements.txt file to the setup.cfg. Is this correct?

`umami` should be installable with standard tooling (e.g. `python -m pip install` or `uv pip install`) without needing any additional files, and so having the runtime requirements be specified in the `setup.cfg` is necessary. Though (going into question 2 a bit) the known lower bounds (`>=`) of the dependencies should be specified (unless there is a known breaking change in a requirement which necessitates an upper bound be added `>=X,<Y` or a pin `==Y`).

> In short, this means that for libraries all the required runtime dependencies need to be included in the packaging metadata with known lower bounds (if you don't want to empirically determine these lower bounds it is okay to select reasonable known working versions) so the installer knows what versions are required. For a Python application, in addition to the library requirements, a `requirements.txt` or (preferably) a lock file (created by some well established tool like `pip-tools`, `conda-lock`, `poetry`, `uv` as there is [currently no formal Python lock file standard](https://discuss.python.org/t/lock-files-again-but-this-time-w-sdists/46593)) should also be provided.

(I'm copying that from https://github.com/umami-hep/umami/issues/6#issue-2181036198 to avoid you having to reread that all again)

> If so, I saw there is a possibility to do this dynamically. I guess this would be good?

Sorry, can you elaborate more on what "do this dynamically" means?

> 2: Which packages do you talk about when writing "correct application run requirements with deploy/run versions pinned"? From my point, I think we pinned down all the different packages that we use with the version required. Could you clarify what you meant with that?

When answering this question it helps to have a clear understanding of what the [difference between a library and an application is](https://iscinumpy.dev/post/app-vs-library/). As it was agreed that `umami` is not intended to be used as a library, but as a Python application, it is then important to _fully specify_ the runtime environment. The contents of the `requirements.txt`

https://github.com/umami-hep/umami/blob/c7e7d164f554f07a20dcfbf760a47c828e638abc/requirements.txt#L1-L16

do not do this as they only specify the versions of the high level requirements &mdash; they do not specify the versions of those requirements' dependencies at all. A practical way to see this it so generate a lock file from the `requirements.txt` with `uv` and note that number of dependencies is quite different (there are 25 dependencies that are not being specified and could very well break any and all behavior in future updates):

```console
$ docker run --rm -ti python:3.12 /bin/bash
root@666372a306f4:/# python -m venv venv && . venv/bin/activate
(venv) root@666372a306f4:/# python -m pip --quiet install --upgrade uv
(venv) root@666372a306f4:/# curl -sLO https://raw.githubusercontent.com/umami-hep/umami/c7e7d164f554f07a20dcfbf760a47c828e638abc/requirements.txt
(venv) root@666372a306f4:/# uv pip compile --generate-hashes --output-file=requirements.lock requirements.txt
(venv) root@666372a306f4:/# grep "==" requirements.txt | wc -l
16
(venv) root@666372a306f4:/# grep "==" requirements.lock | wc -l
41
```

An application should have the runtime environment fully specified with a lock file built from a high level summary of the direct retirements. The existing `requirements.txt` is an example of such a high level specification.

---

If you have additional questions please feel free to ask for clarification here. I'm happy to discuss anything. As noted in email communication I'm currently a bit occupied for personal/professional reasons, but with this now resurfaced in GitHub Issues I can be substantially more responsive to help you finish this JOSS submission.

**Comment by afroch** [comment]
Hi @matthewfeickert. Thanks for the very nice answer! I tried to adapt all the comments you had and cumulated them in the following MR (Gitlab) [!752](https://gitlab.cern.ch/atlas-flavor-tagging-tools/algorithms/umami/-/merge_requests/752). The project should be visible to everyone and once it is merged into master, the github instance of umami will be updated by a cron job.

Feel free to drop any comments on the MR or here.

**Comment by matthewfeickert** [comment]
(I'll double post here and on CERN's GitLab)

@afroch MR https://gitlab.cern.ch/atlas-flavor-tagging-tools/algorithms/umami/-/merge_requests/752 is sitll in draft, so maybe it isn't done yet, but from https://github.com/umami-hep/umami/issues/6#issuecomment-2354707896 you still need to define the runtime requirements in the metadata (this is `install_requires` for `setuptools` and `dependencies` for all other build backends).

Aside: (This is a comment NOT a requirement to proceed) You don't use any `setuptools` particular features here, so you could also migrate this all to `hatchling` and simplify everything to exist in `pyproject.toml`. `hatch` does this for you automatically with `hatch new --init`

```console
$ docker run --rm -ti python:3.12 /bin/bash
root@88dde1034a68:/# python -m venv venv && . venv/bin/activate
(venv) root@88dde1034a68:/# git clone --recursive https://github.com/umami-hep/umami.git
Cloning into 'umami'...
remote: Enumerating objects: 11736, done.
remote: Counting objects: 100% (829/829), done.
remote: Compressing objects: 100% (386/386), done.
remote: Total 11736 (delta 466), reused 792 (delta 433), pack-reused 10907 (from 1)
Receiving objects: 100% (11736/11736), 93.69 MiB | 46.17 MiB/s, done.
Resolving deltas: 100% (7807/7807), done.
(venv) root@88dde1034a68:/# cd umami/
(venv) root@88dde1034a68:/umami# python -m pip --quiet install --upgrade hatch
(venv) root@88dde1034a68:/umami# hatch new --init
(venv) root@88dde1034a68:/umami# git diff
diff --git a/pyproject.toml b/pyproject.toml
index 11d1e97..caadc7b 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,5 +1,31 @@
 [build-system]
-requires = ["setuptools", "wheel"]
+requires = ["hatchling"]
+build-backend = "hatchling.build"
+
+[project]
+name = "umami"
+dynamic = ["version"]
+description = "Machine learning based flavour tagging training framework."
+readme = "README.md"
+license = ""
+requires-python = ">= 3.8"
+keywords = [
+    "flavour",
+    "learning",
+    "machine",
+    "tagging",
+]
+
+[project.urls]
+Homepage = "https://gitlab.cern.ch/atlas-flavor-tagging-tools/algorithms/umami"
+
+[tool.hatch.version]
+path = "umami/__init__.py"
+
+[tool.hatch.build.targets.sdist]
+include = [
+    "/umami",
+]
 
 [tool.black]
 line-length = 88
```

Also do not try to make `requirements.txt` a self created lock file. This is an anti-goal and now removes any benefits of understanding what your high level dependencies are. Keep a high level runtime requirements file separate from a **tool** built lock file. If the project runtime requirements are properly defined you can generate a lock file from them like in the following example:

```console
$ docker run --rm -ti python:3.12 /bin/bash
root@49d9431484e0:/# python -m venv venv && . venv/bin/activate
(venv) root@49d9431484e0:/# python -m pip --quiet install --upgrade uv
(venv) root@49d9431484e0:/# git clone --recursive https://github.com/scikit-hep/awkward.git
(venv) root@49d9431484e0:/# cd awkward/
(venv) root@49d9431484e0:/# uv pip compile --generate-hashes --output-file=requirements.txt pyproject.toml
```

This example demos how you could keep the `requirements.txt` name without having to teach people about `requirements.lock` and also benefit from having your requirements defined properly.

**Comment by afroch** [comment]
Hi @matthewfeickert. The corresponding MR adapting all your suggestions ([!752](https://gitlab.cern.ch/atlas-flavor-tagging-tools/algorithms/umami/-/merge_requests/752) in Gitlab) is now merged and we fixed all the outstanding issues with the deployment part of the pipeline.

**Comment by matthewfeickert** [comment]
Thanks @afroch. :+1:
